### PR TITLE
MPS dev build

### PR DESCRIPTION
**Pull request by ngam**
TODO:
- [ ] check if macos-12 is needed for vmImage...
- [x] see if full matrix is generated correctly... 
- [ ] add strict run constraints about 12.3
---

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
Currently this fails because it of the sdk 11 from conda-forge...

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
@hmaarrfk this works fine. I downloaded one of these artifacts and uploaded them to my channel: anaconda.org/ngam. Things look good. 

<details>

```

(py39) ~$ mamba create -n pt39 'pytorch==1.13.0=*mps*' -c ngam -c conda-forge

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.23.3) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['pytorch==1.13.0[build=*mps*]']

ngam/noarch                                        130.0 B @ 465.0 B/s  0.3s
ngam/osx-arm64                                     699.0 B @   2.2kB/s  0.3s
conda-forge/osx-arm64                                2.5MB @   2.6MB/s  1.0s
conda-forge/noarch                                   8.3MB @   3.4MB/s  2.7s
Transaction

  Prefix: /Users/ngam/.Mambaforge-MacOSX-arm64/envs/pt39

  Updating specs:

   - pytorch==1.13.0[build=*mps*]


  Package                  Version  Build                 Channel                     Size
────────────────────────────────────────────────────────────────────────────────────────────
  Install:
────────────────────────────────────────────────────────────────────────────────────────────

  + bzip2                    1.0.8  h3422bc3_4            conda-forge/osx-arm64     Cached
  + ca-certificates    2022.5.18.1  h4653dfc_0            conda-forge/osx-arm64     Cached
  + cffi                    1.15.0  py39h52b1de0_0        conda-forge/osx-arm64     Cached
  + libblas                  3.9.0  14_osxarm64_openblas  conda-forge/osx-arm64       13kB
  + libcblas                 3.9.0  14_osxarm64_openblas  conda-forge/osx-arm64       13kB
  + libcxx                  14.0.4  h6a5c8ee_0            conda-forge/osx-arm64     Cached
  + libffi                   3.4.2  h3422bc3_5            conda-forge/osx-arm64     Cached
  + libgfortran         5.0.0.dev0  11_0_1_hf114ba7_23    conda-forge/osx-arm64       20kB
  + libgfortran5       11.0.1.dev0  hf114ba7_23           conda-forge/osx-arm64      923kB
  + liblapack                3.9.0  14_osxarm64_openblas  conda-forge/osx-arm64       13kB
  + libopenblas             0.3.20  openmp_h2209c59_0     conda-forge/osx-arm64     Cached
  + libprotobuf             3.20.1  h332123e_0            conda-forge/osx-arm64     Cached
  + libzlib                 1.2.12  h90dfc92_0            conda-forge/osx-arm64     Cached
  + llvm-openmp             14.0.4  hd125106_0            conda-forge/osx-arm64     Cached
  + ncurses                    6.3  h07bb92c_1            conda-forge/osx-arm64     Cached
  + ninja                   1.11.0  hf86a087_0            conda-forge/osx-arm64     Cached
  + numpy                   1.22.4  py39h7df2422_0        conda-forge/osx-arm64     Cached
  + openssl                  3.0.3  ha287fd2_0            conda-forge/osx-arm64        2MB
  + pip                     22.1.1  pyhd8ed1ab_0          conda-forge/noarch        Cached
  + pycparser                 2.21  pyhd8ed1ab_0          conda-forge/noarch        Cached
  + python                  3.9.13  h96fcbfb_0_cpython    conda-forge/osx-arm64       13MB
  + python_abi                 3.9  2_cp39                conda-forge/osx-arm64        4kB
  + pytorch                 1.13.0  mps_py39he80704b_1    ngam/osx-arm64              41MB
  + readline                   8.1  hedafd6a_0            conda-forge/osx-arm64     Cached
  + setuptools              62.3.2  py39h2804cbe_0        conda-forge/osx-arm64     Cached
  + sleef                    3.5.1  h156473d_2            conda-forge/osx-arm64     Cached
  + sqlite                  3.38.5  h40dfcc0_0            conda-forge/osx-arm64     Cached
  + tk                      8.6.12  he1e0b03_0            conda-forge/osx-arm64     Cached
  + typing_extensions        4.2.0  pyha770c72_1          conda-forge/noarch        Cached
  + tzdata                   2022a  h191b570_0            conda-forge/noarch        Cached
  + wheel                   0.37.1  pyhd8ed1ab_0          conda-forge/noarch        Cached
  + xz                       5.2.5  h642e427_1            conda-forge/osx-arm64     Cached
  + zlib                    1.2.12  h90dfc92_0            conda-forge/osx-arm64     Cached

  Summary:

  Install: 33 packages

  Total download: 57MB

────────────────────────────────────────────────────────────────────────────────────────────

Confirm changes: [Y/n] y
libcblas                                            12.7kB @  51.8kB/s  0.2s
libblas                                             12.8kB @  47.4kB/s  0.3s
libgfortran5                                       922.6kB @   3.3MB/s  0.3s
libgfortran                                         19.7kB @  71.3kB/s  0.3s
python_abi                                           4.1kB @  11.7kB/s  0.1s
liblapack                                           12.8kB @  36.1kB/s  0.1s
openssl                                              2.5MB @   4.9MB/s  0.5s
python                                              12.8MB @  10.0MB/s  1.0s
pytorch                                             40.5MB @   6.3MB/s  6.2s
Preparing transaction: done
Verifying transaction: done
Executing transaction: done

To activate this environment, use

     $ mamba activate pt39

To deactivate an active environment, use

     $ mamba deactivate

(py39) ~$ mamba activate pt39
(pt39) ~$ python3
Python 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00) 
[Clang 13.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.backends.mps.is_available()
True
>>> 

```

</details>

The problem is how should we incorporate this going forward? I think it is easiest to just make an mps branch and manually adjust it at every iteration. That's because this is a huge jump in terms of minimum sdk req (sdk 12.3, fwiw sdk 12.4 hasn't even been released yet, and macos 12.4 was just released like a week or so ago). It will also likely disrupt the rhythm of dealing with the cuda builds, etc. 

If we have make an mps branch, I can do all the manual work and submit PRs targeting that branch. The good news is that this will pass under the CI limits. 

In summary, I propose:
1. we follow our pattern of naming, we call these builds *mps* (compared to cuda and cpu)
2. we house these builds in an mps branch (I can then submit PRs targetting that branch with only mps builds that will go to the main channel on anaconda.org)
3. I can start tweaking this to make it fully compatible with Apple's ecosystem (e.g. targeting metal backend as well (for mobile mostly) and targeting accelerate BLAS implementation, etc.) --- in other words, fully optimizing this for Apple's *newer* hardware. This will take time and a lot of tweaks that will not be easily extended to other builds, so reinforcing housing this in a separate branch
4. All current implementations (cpu + cuda) remain the same going forward in the main branch.

**Comment by ngam** [comment]
(no rush btw, an actual release is likely at least few weeks away)

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -1,2 +1,11 @@
+osx_impl:  # [osx]
+  - mps    # [osx]
+  - nomps  # [osx]
+
+MACOSX_DEPLOYMENT_TARGET:  # [osx]
+  - "12.3"                 # [osx]
+MACOSX_SDK_VERSION:        # [osx]
+  - "12.3"                 # [osx]
+
```
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:9]:
```suggestion
MACOSX_DEPLOYMENT_TARGET:  # [osx]
  - 11.0                   # [osx and arm64]
  - 10.9                   # [osx and x86_64]
  - 12.3                   # [osx]
MACOSX_SDK_VERSION:        # [osx]
  - 11.0                   # [osx and arm64]
  - 10.9                   # [osx and x86_64]
  - 12.3                   # [osx]

zip_keys:                           # [osx]
  - - osx_impl                      # [osx]
    - MACOSX_DEPLOYMENT_TARGET      # [osx]
    - MACOSX_SDK_VERSION            # [osx]
```

can you try with this and rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -7,19 +7,20 @@ package:
...
-  skip: true  # [win]
+  skip: true  # [not osx]
 
 outputs:
   - name: pytorch
     build:
       string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
-      string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None"]
+      string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None" and osx_impl != "mps"]
+      string: mps_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None" and osx_impl == "mps"]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:23]: I think there is a similar build number below. You might want to add an other `pytorch-cpu` package for MAC and just pin it to the MPS package.

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml (hunk 2 above)
Comment (ngam) [reply recipe/meta.yaml:23]: Updated, it's a bit of matrix now... 😰

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

File: recipe/conda_build_config.yaml [hunk 3]
```diff
@@ -1,2 +1,19 @@
+osx_impl:  # [osx]
+  - nomps  # [osx]
+  - mps    # [osx]
+
+MACOSX_DEPLOYMENT_TARGET:  # [osx]
+  - 11.0                   # [osx and arm64]
+  - 10.9                   # [osx and x86_64]
+  - 12.3                   # [osx]
+MACOSX_SDK_VERSION:        # [osx]
```
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:9]: it may be that we don't need to specify this.

Comment (ngam) [reply recipe/conda_build_config.yaml:9]: No, it is working as desired

Comment (ngam) [reply recipe/conda_build_config.yaml:9]: the current problem is with my logic selectors in the names 😵‍💫

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

File: recipe/conda_build_config.yaml (hunk 3 above)
Comment (ngam) [reply recipe/conda_build_config.yaml:9]: I think I know what you mean tho, the 11.0 and 10.9 are the default values, but how can we zip them together without being explicit?

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:9]: I have a feeling that `MACOSX_SDK_VERSION` will be specified to the other defined if not specified.

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:9]: maybe.

Comment (ngam) [reply recipe/conda_build_config.yaml:9]: will test locally and report back

Comment (ngam) [reply recipe/conda_build_config.yaml:9]:
```
  zip fields in zip_key group frozenset({'osx_impl', 'MACOSX_DEPLOYMENT_TARGET', 'MACOSX_SDK_VERSION'}) are not all the same length

```

**Comment by ngam** [comment]
NOMPS failures:

```
  /Users/runner/miniforge3/conda-bld/pytorch-recipe_1653768950821/work/third_party/XNNPACK/src/jit/memory.c:35:68: error: use of undeclared identifier 'MAP_ANONYMOUS'
    void* p = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
                                                                     ^
  1 error generated.

```

I will check upstream to see their minimum SDK. This looks like an SDK error (it worked completely fine with 12.3). I believe they use sdk==11 as a minimum, but I will check/confirm

**Comment by hmaarrfk** [comment]
it might be that you need a version bump.

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
blocked by https://github.com/conda-forge/conda-forge-ci-setup-feedstock/issues/200

File: recipe/conda_build_config.yaml (hunk 3 above)
Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:9]: remove `MACOSX_SDK_VERSION` from the zip keys below as well.

**Comment by ngam** [comment]
@conda-forge-admin, please rerender

**Comment by ngam** [comment]
I have a feeling that https://github.com/conda-forge/conda-forge-ci-setup-feedstock/issues/200 won't really fix this. My recommendation is to have an mps branch for the mps builds --- then everything else can be skipped and no changed to the main workflow are needed...
